### PR TITLE
install gecko-dev as submodule and add scripts to compile to wasm

### PR DESCRIPTION
**Pull request by JakeChampion**
This inlines spidermonkey-wasi-embedding so that changes from our gecko-dev only need to be pulled into one repository and not two

### CODE REVIEW

File: runtime/spidermonkey/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+Scripts for building and downloading pre-built versions of SpiderMonkey, compiled to wasm32-wasi as a static library.
+
+## Building from upstream source
+The `build-engine.sh` script can be used to build either release or debug builds. It's recommended that this script only be used in CI environments, as it bootstraps a build environment for SpiderMonkey on each invokation.
+
+The source is retrieved from the repository in `gecko-repository` at the revision in `gecko-revision`.
```
Comment (JakeChampion) [inline runtime/spidermonkey/README.md:6]:
```suggestion
```

File: runtime/spidermonkey/README.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
+Scripts for building and downloading pre-built versions of SpiderMonkey, compiled to wasm32-wasi as a static library.
```
Comment (JakeChampion) [inline runtime/spidermonkey/README.md:1]:
```suggestion
Scripts for building SpiderMonkey, compiled to wasm32-wasi as a static library.
```

File: runtime/spidermonkey/README.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+- `include`, containing the public header files
+- `lib`, containing the object files and static libraries needed to statically link SpiderMonkey
+- `obj`, the output directory the SpiderMonkey build system creates. This can be ignored or deleted
+
+## Downloading pre-built releases
+The `download-engine.sh` script can be used to download pre-built object files and headers for static linking for either release or debug builds. The object files are compiled to wasm32-wasi, and can be used on any platform.
+
+Running `download-engine.sh [debug]` will download a tarball for a SpiderMonkey build (release by default, debug if the `debug` argument is passed), and extract it. When successful, it'll result in the subdirectories `lib` and `include` in the current working directory.
+
+**Note:** `download-engine.sh` can be called from anywhere, but must itself reside in a git checkout whose `origin` is the github repository to download the release from. It expects that repository to have a release whose name is derived from the current git revision of the containing checkout.
```
Comment (JakeChampion) [inline runtime/spidermonkey/README.md:24]:
```suggestion
```

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -69,9 +69,7 @@ jobs:
           runtime/spidermonkey/${{ matrix.profile }}
         key: cache-${{ hashFiles(
             'runtime/spidermonkey/build-engine.sh',
-            'runtime/spidermonkey/gecko-revision',
-            'runtime/spidermonkey/object-files.list',
-            'runtime/rust-toolchain.toml'
+            'runtime/spidermonkey/object-files.list'
```
Comment (elliottt) [inline .github/workflows/main.yml:72]: Won't this use cached builds for commits that change the gecko-dev submodule?

File: runtime/spidermonkey/LICENSE [hunk 5]
```diff
@@ -0,0 +1,220 @@
+
+                                 Apache License
+                           Version 2.0, January 2004
```
Comment (elliottt) [inline runtime/spidermonkey/LICENSE:3]: Should we delete this file, as the top-level license is the same?

**Review by elliottt — COMMENTED** [review]
Looks good! Just two questions before we merge.

File: .github/workflows/main.yml (hunk 4 above)
Comment (JakeChampion) [reply .github/workflows/main.yml:72]: silly me, yes you are correct, shall we add the entire gecko-dev folder?

Comment (elliottt) [reply .github/workflows/main.yml]: Would that consider the submodule hash, or the contents of gecko-dev? I think the submodule hash is what we want, so maybe that just works out :)

Comment (JakeChampion) [reply .github/workflows/main.yml]: I have added the commit hash of gecko-dev into the cache key [`a1dab90` (#636)](https://github.com/fastly/js-compute-runtime/pull/636/commits/a1dab90528f175617c02bb0964fd6fca79677295)

Comment (elliottt) [reply .github/workflows/main.yml]: Oh, you did that in your last commit already, thanks!

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -61,6 +61,12 @@ jobs:
       with:
         path: runtime
 
+    - name: gecko-dev hash
+      id: gecko-dev-hash
+      run: cd runtime/spidermonkey/gecko-dev && echo "GECKO_DEV_HASH=$(git rev-parse HEAD) >> "$GITHUB_OUTPUT"
```
Comment (elliottt) [inline .github/workflows/main.yml]:
```suggestion
      run: cd runtime/spidermonkey/gecko-dev && echo "GECKO_DEV_HASH=$(git rev-parse HEAD)" >> "$GITHUB_OUTPUT"
```

File: .github/workflows/main.yml [hunk 7]
```diff
@@ -69,10 +75,8 @@ jobs:
           runtime/spidermonkey/${{ matrix.profile }}
         key: cache-${{ hashFiles(
             'runtime/spidermonkey/build-engine.sh',
-            'runtime/spidermonkey/gecko-revision',
-            'runtime/spidermonkey/object-files.list',
-            'runtime/rust-toolchain.toml'
-          ) }}-${{ matrix.profile }}
+            'runtime/spidermonkey/object-files.list'
+          ) }}-${{ matrix.profile }}-${{ steps.gecko-dev-hash.outputs.GECKO_DEV_HASH }}
```
Comment (elliottt) [inline .github/workflows/main.yml:79]: :tada:

**Review by elliottt — APPROVED** [review]
This looks great to me! I'm really looking forward to a shorter development cycle with the intermediate repo gone :D

**Pull request merged**